week5/Q1.py

```diff
@@ -0,0 +1,149 @@
+class OrientedGraph:
+	'''
+	Implementation of oriented graph. All of the nodes have unique values.
+
+	Fields:
+		nodes: set of Nodes, all nodes of the graph
+	'''
+	def __init__(self):
+		self.nodes = set()
+
+	def add(self, value, parent_value = None):
+		'''
+		Adds new Node to the OrientedGraph or updates existing ones.
+
+		Args:
+			value: str
+			parent_value: str
+		'''
+		node = None
+		parent = None
+		for inode in self.nodes:
+			if inode.value == value:
+				node = inode
+			if inode.value == parent_value:
+				parent = inode
+
+		if parent_value:
+			if node:
+				node.parents = node.parents.union({parent})
+				parent.children = parent.children.union({node})
+			else:
+				node = Node(value, parents = {parent})
```

Comment:
You can probably merge the two sides of the `if`-statement somewhat.
If the `node` does not exist, you can create it and then how you insert it in the graph is essentially the same in both cases.